File: android/src/main/kotlin/com/neiry/neiry_kit/MemsBridge.kt
package com.neiry.neiry_kit

import io.flutter.plugin.common.EventChannel

/**
 * Kotlin bridge for the `clCMEMS` classifier. Owns the native MEMS handle
 * and exposes factory methods (plain and calibrated) plus a stream handler for
 * the memsData event channel.
 *
 * Mirrors [MemsBridge.swift] in behavior.
 */
class MemsBridge(private val nativeBridge: NativeBridge) {

    private var handle: Long = 0L

    // ── Stream handlers ───────────────────────────────────────────────────────

    inner class MemsStreamHandler(
        private val setSink: (EventChannel.EventSink?) -> Unit
    ) : EventChannel.StreamHandler {
        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
            setSink(events)
        }
        override fun onCancel(arguments: Any?) {
            setSink(null)
        }
    }

    val memsDataHandler = MemsStreamHandler { nativeBridge.nativeSetMemsDataSink(it) }

    /** Returns all (channelId, handler) pairs for EventChannel registration in the plugin. */
    fun allStreamHandlers(): List<Pair<String, EventChannel.StreamHandler>> = listOf(
        "neiry_kit/events/memsData" to memsDataHandler,
    )

    // ── Factory methods ───────────────────────────────────────────────────────

    /** Creates an uncalibrated MEMS classifier for the given device handle. */
    fun create(deviceHandle: Long) {
        if (handle != 0L) dispose()
        try {
            handle = nativeBridge.nativeCreateMems(deviceHandle)
        } catch (e: RuntimeException) {
            throw parseSdkError(e.message ?: "255|Unknown error")
        }
    }

    /**
     * Creates a calibrated MEMS classifier for the given device handle.
     * If [calibrationData] is non-null, its fields are imported into the
     * calibrator before creating the classifier.
     */
    fun createCalibrated(deviceHandle: Long, calibrationData: Map<String, Any>?) {
        if (handle != 0L) dispose()
        val hasData = calibrationData != null
        val ts                                 = (calibrationData?.get("ts") as? Number)?.toLong()    ?: 0L
        val failReason                         = (calibrationData?.get("failReason") as? Number)?.toInt()  ?: 0
        val individualFrequency                = (calibrationData?.get("individualFrequency") as? Number)?.toFloat()             ?: 0f
        val individualPeakFrequency            = (calibrationData?.get("individualPeakFrequency") as? Number)?.toFloat()         ?: 0f
        val individualPeakFrequencyPower       = (calibrationData?.get("individualPeakFrequencyPower") as? Number)?.toFloat()    ?: 0f
        val individualPeakFrequencySuppression = (calibrationData?.get("individualPeakFrequencySuppression") as? Number)?.toFloat() ?: 0f
        val individualBandwidth                = (calibrationData?.get("individualBandwidth") as? Number)?.toFloat()             ?: 0f
        val individualNormalizedPower          = (calibrationData?.get("individualNormalizedPower") as? Number)?.toFloat()       ?: 0f
        val lowerFrequency                     = (calibrationData?.get("lowerFrequency") as? Number)?.toFloat()                  ?: 0f
        val upperFrequency                     = (calibrationData?.get("upperFrequency") as? Number)?.toFloat()                  ?: 0f

        try {
            handle = nativeBridge.nativeCreateMemsCalibrated(
                deviceHandle,
                hasData,
                ts,
                failReason,
                individualFrequency,
                individualPeakFrequency,
                individualPeakFrequencyPower,
                individualPeakFrequencySuppression,
                individualBandwidth,
                individualNormalizedPower,
                lowerFrequency,
                upperFrequency,
            )
        } catch (e: RuntimeException) {
            throw parseSdkError(e.message ?: "255|Unknown error")
        }
    }

    /** Unregisters callbacks and clears the MEMS handle. */
    fun dispose() {
        if (handle != 0L) {
            nativeBridge.nativeDisposeMems(handle)
            handle = 0L
        }
    }
}

File: ios/Classes/classifiers/MemsBridge.swift
import Flutter
import UIKit

// MARK: - MemsBridge

class MemsBridge: NSObject {

    // MARK: Static state (used by C callbacks which cannot capture `self`)

    /// The currently active bridge. C callbacks reach the instance through this
    /// static weak reference because `clCMEMS_Set*Event` callbacks have no
    /// `void* context` parameter.
    private static weak var activeBridge: MemsBridge?

    // MARK: Stream handlers

    let memsDataHandler = DeviceStreamHandler(channelId: "neiry_kit/events/memsData")

    /// Returns all (channelId, handler) pairs for EventChannel registration in the plugin.
    func allStreamHandlers() -> [(String, FlutterStreamHandler)] {
        return [
            (memsDataHandler.channelId, memsDataHandler),
        ]
    }

    // MARK: Instance state

    private var mems: OpaquePointer?

    // MARK: - Public API

    /// Creates an uncalibrated MEMS classifier for the given device handle.
    func create(device: OpaquePointer) throws {
        if mems != nil { unregisterCallbacks() }
        var error = clCError()
        mems = clCMEMS_Create(device, &error)
        try checkCError(error)
        try registerCallbacks()
    }

    /// Creates a calibrated MEMS classifier for the given device handle.
    /// If `calibrationData` is provided, it is imported into the calibrator before creating.
    func createCalibrated(device: OpaquePointer, calibrationData: [String: Any]?) throws {
        if mems != nil { unregisterCallbacks() }
        guard let calibrator = clCNFBCalibrator_CreateOrGet(device) else {
            throw FlutterError(code: "NULL_HANDLE", message: "clCNFBCalibrator_CreateOrGet returned nil", details: nil)
        }
        if let calibrationData = calibrationData {
            var data = clCIndividualNFBData()
            if let ts = calibrationData["ts"] as? Int {
                data.timestampMilli = Int64(ts)
            }
            if let failReasonRaw = calibrationData["failReason"] as? Int {
                data.failReason = clCIndividualNFBCalibrationFailReason(rawValue: UInt32(failReasonRaw))
            }
            if let v = calibrationData["individualFrequency"] as? Double { data.individualFrequency = Float(v) }
            if let v = calibrationData["individualPeakFrequency"] as? Double { data.individualPeakFrequency = Float(v) }
            if let v = calibrationData["individualPeakFrequencyPower"] as? Double { data.individualPeakFrequencyPower = Float(v) }
            if let v = calibrationData["individualPeakFrequencySuppression"] as? Double { data.individualPeakFrequencySuppression = Float(v) }
            if let v = calibrationData["individualBandwidth"] as? Double { data.individualBandwidth = Float(v) }
            if let v = calibrationData["individualNormalizedPower"] as? Double { data.individualNormalizedPower = Float(v) }
            if let v = calibrationData["lowerFrequency"] as? Double { data.lowerFrequency = Float(v) }
            if let v = calibrationData["upperFrequency"] as? Double { data.upperFrequency = Float(v) }
            var importError = clCError()
            clCNFBCalibrator_ImportIndividualNFBData(calibrator, &data, &importError)
            try checkCError(importError)
        }
        var error = clCError()
        mems = clCMEMS_CreateCalibrated(device, calibrator, &error)
        try checkCError(error)
        try registerCallbacks()
    }

    /// Unregisters all callbacks and releases the MEMS handle.
    /// There is no `clCMEMS_Destroy` in the SDK — disposal is callback-only.
    func dispose() {
        unregisterCallbacks()
        mems = nil
    }

    // MARK: - Callbacks

    private func registerCallbacks() throws {
        guard let mems = mems else { return }
        MemsBridge.activeBridge = self

        var e1 = clCError()
        clCMEMS_SetOnMEMSTimedDataUpdateEvent(mems, { _, memsData in
            guard let bridge = MemsBridge.activeBridge,
                  let memsData = memsData else { return }
            let count = clCMEMSTimedData_GetCount(memsData)
            var samples: [[String: Any]] = []
            for i in 0..<count {
                let accel = clCMEMSTimedData_GetAccelerometer(memsData, i)
                let gyro = clCMEMSTimedData_GetGyroscope(memsData, i)
                let ts = clCMEMSTimedData_GetTimestampMilli(memsData, i)
                let sample: [String: Any] = [
                    "ax": accel.x,
                    "ay": accel.y,
                    "az": accel.z,
                    "gx": gyro.x,
                    "gy": gyro.y,
                    "gz": gyro.z,
                    "ts": ts,
                ]
                samples.append(sample)
            }
            bridge.memsDataHandler.sendList(samples)
        }, &e1)
        try checkCError(e1)
    }

    private func unregisterCallbacks() {
        guard let mems = mems else { return }
        var e = clCError()
        clCMEMS_SetOnMEMSTimedDataUpdateEvent(mems, nil, &e)
        if MemsBridge.activeBridge === self {
            MemsBridge.activeBridge = nil
        }
    }
}
